### PR TITLE
Shard management and searching with HMAC keying 

### PR DESCRIPTION
**Pull request by hoytak**
Ready for Review: 

- Adds hmac keys to the shard format and shard management infrastructure.
- Adds function to do a streaming rebuild of a shard with hmac keys applied.

### CODE REVIEW

File: mdb_shard/src/shard_file_handle.rs [hunk 1]
```diff
@@ -130,6 +132,33 @@ impl MDBShardFile {
         Ok(ret)
     }
 
+    /// Write out the current shard, re-keyed with an hmac key, to the output directory in question, returning
+    /// the foll path to the new shard.
```
Comment (seanses) [inline mdb_shard/src/shard_file_handle.rs]: typo
```suggestion
    /// the full path to the new shard.
```

File: mdb_shard/src/shard_format.rs [hunk 2]
```diff
@@ -144,9 +146,9 @@ impl MDBShardFileFooter {
 
         // Do a version check as a simple guard against using this in an old repository
         if version != MDB_SHARD_FOOTER_VERSION {
-            return Err(MDBShardError::ShardVersionError(format!(
-                "Error: Expected header version {MDB_SHARD_FOOTER_VERSION}, got {version}"
-            )));
+            //return Err(MDBShardError::ShardVersionError(
```
Comment (seanses) [inline mdb_shard/src/shard_format.rs]: we can drop this commented snippet

File: mdb_shard/src/shard_format.rs [hunk 3]
```diff
@@ -643,14 +668,22 @@ impl MDBShardInfo {
...
+        unkeyed_query_hashes: &[MerkleHash],
         cas_entry_index: u32,
         cas_chunk_offset: u32,
     ) -> Result<Option<(usize, FileDataSequenceEntry)>> {
-        if query_hashes.is_empty() {
+        if unkeyed_query_hashes.is_empty() {
             return Ok(None);
         }
 
+        let keyed_hash = |h: MerkleHash| {
```
Comment (seanses) [inline mdb_shard/src/shard_format.rs]: Is this identical to the above `pub fn keyed_chunk_hash`?

Comment (hoytak) [reply mdb_shard/src/shard_format.rs]: Yes, will consolidate.

File: mdb_shard/src/shard_format.rs (hunk 2 above)
Comment (hoytak) [reply mdb_shard/src/shard_format.rs]: The panic below was for catching an annoying error during debugging, reverted that.

**Pull request merged**